### PR TITLE
Add RMI support

### PR DESCRIPTION
**Pull request by pawelchcki**
This PR in its current form adds instrumentation of client and server endpoints used in Java RMI communication.
Both new RMI dynamically generated communication stubs and `rmic` generated Stubs are supported.

```
 * Main entry point for transferring context between RMI service.
 *
 * <p>We into StreamRemoteCall constructor used for invoking remote tasks and performs a
 * backwards compatible check to ensure if the other side is prepared to receive context propagation
 * messages then if successful sends a context propagation message
 *
 * <p>Context propagation consist of a Serialized HashMap with all data set by usual context
 * injection, which includes things like sampling priority, trace and parent id
 *
 * <p>As well as optional baggage items
 *
 * <p>On the other side of the communication a special Dispatcher is created when a message with
 * DD_CONTEXT_CALL_ID is received.
 *
 * <p>If the server is not instrumented first call will gracefully fail just like any other unknown
 * call. With small caveat that this first call needs to *not* have any parameters, since those will
 * not be read from connection and instead will be interpreted as another remote instruction, but
 * that instruction will essentially be garbage data and will cause the parsing loop to throw
 * exception and shutdown the connection which we do not want
```

### CODE REVIEW

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ObjectTableAdvice.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+@Slf4j
+public class ObjectTableAdvice {
+  public static final DummyRemote DUMMY_REMOTE = new DummyRemote();
+
+  @Advice.OnMethodExit(suppress = Throwable.class)
+  public static void methodExit(
+      @Advice.Argument(0) final Object oe, @Advice.Return(readOnly = false) Target result) {
+
+    if (!DD_CONTEXT_CALL_ID.toString().equals(oe.toString())) {
```
Comment (summner) [inline dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ObjectTableAdvice.java]: While I'm not super happy about using toString in if in this case it avoids using reflection. Its also easily testable for all supported JDKs and is not supposed to change between patch releases.

Hmm actually I will add a comment explaining what object we actually are comparing here.

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/server/RmiServerInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  public static class ServerAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class, inline = true)
+    public static AgentScope onEnter(
+        @Advice.This final Object thiz, @Advice.Origin(value = "#m") final String method) {
+      final ContextStore<Thread, AgentSpan.Context> callableContextStore =
+          InstrumentationContext.get(Thread.class, AgentSpan.Context.class);
+
+      final AgentSpan span =
+          startSpan(callableContextStore)
+              .setTag(DDTags.RESOURCE_NAME, thiz.getClass().getSimpleName() + "#" + method)
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/server/RmiServerInstrumentation.java]: Use decorator helper method here too.

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/server/RmiServerInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    public static AgentSpan startSpan(
+        final ContextStore<Thread, AgentSpan.Context> callableContextStore) {
+      if (activeSpan() != null) {
+        return AgentTracer.startSpan("rmi.request");
+      }
+
+      final AgentSpan.Context context = callableContextStore.get(Thread.currentThread());
+
+      if (context == null) {
+        return AgentTracer.startSpan("rmi.request");
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/server/RmiServerInstrumentation.java]: AgentTracer methods were designed for static import.

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/server/ServerDecorator.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package datadog.trace.instrumentation.rmi.server;
+
+import datadog.trace.agent.decorator.BaseDecorator;
+import datadog.trace.api.DDSpanTypes;
+
+public class ServerDecorator extends BaseDecorator {
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/server/ServerDecorator.java]: Rename to `RmiServerDecorator` and extend `ServerDecorator`.

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ObjectTableAdvice.java (hunk 1 above)
Comment (pawelchcki) [reply dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ObjectTableAdvice.java]: Whoops wrong account :)

[the remark this answers is not in the document]
Comment (pawelchcki) [reply dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ObjectTableAdvice.java]: oops forgot to push a commit with the comment - added now

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ContextDispatcher.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    final ObjectInput in = call.getInputStream();
+    final int operationId = in.readInt();
+    in.readLong(); // skip 8 bytes
+
+    if (operationId == StreamRemoteCallConstructorAdvice.CONTEXT_PASS_OPERATION_ID) {
+      final ContextPayload payload = ContextPayload.read(in);
+      if (payload != null) {
+        final AgentSpan.Context context = propagate().extract(payload, GETTER);
+
+        callableContextStore.put(Thread.currentThread(), context);
```
[the remark this answers is not in the document]
Comment (pawelchcki) [reply dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ContextDispatcher.java]: If I use any of class injected by instrumentation there will be two - or more, instances of those classes.

When instrumenting both server instrumentation and context receiving call sites one will see instance 1, second will see instance 2. 

I think its because the RMI - soon before executing server implementation class sets 
Context class loader on current thread in (`Transport.java`):
```
    ClassLoader ccl = target.getContextClassLoader();

                ClassLoader savedCcl = Thread.currentThread().getContextClassLoader();

                try {
                    setContextClassLoader(ccl);
```

**Comment by pawelchcki** [comment]
Thanks @tylerbenson for the review! Addressed all your comments. With this "thread context local" problem I'm open to alternative solutions since I too don't like the concept of injecting field into Thread object too much.

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ObjectTableAdvice.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+      return;
+    }
+
+    final ContextStore<Thread, AgentSpan.Context> callableContextStore =
+        InstrumentationContext.get(Thread.class, AgentSpan.Context.class);
```
[the remark this answers is not in the document]
Comment (pawelchcki) [reply dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ObjectTableAdvice.java]: see: https://github.com/DataDog/dd-trace-java/pull/1128#discussion_r360355975

It could be possible but might require changes in the agent class injection

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/StreamRemoteCallConstructorAdvice.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+ * not be read from connection and instead will be interpreted as another remote instruction, but
+ * that instruction will essentially be garbage data and will cause the parsing loop to throw
+ * exception and shutdown the connection which we do not want
+ */
+@Slf4j
+public class StreamRemoteCallConstructorAdvice {
+  public static final ObjID ACTIVATOR_ID = new ObjID(ObjID.ACTIVATOR_ID);
+  public static final ObjID DGC_ID = new ObjID(ObjID.DGC_ID);
+  public static final ObjID REGISTRY_ID = new ObjID(ObjID.REGISTRY_ID);
+  public static final ObjID DD_CONTEXT_CALL_ID = new ObjID("Datadog.context_call.v2".hashCode());
```
[the remark this answers is not in the document]
Comment (pawelchcki) [reply dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/StreamRemoteCallConstructorAdvice.java]: there used to be v1 that i shipped to RMI. Just bumped it to 2. Now I reverted it to v1

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/RmiContextInstrumentation.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  @Override
+  public ElementMatcher<? super TypeDescription> typeMatcher() {
+    return not(isInterface())
+        .and(
+            safeHasSuperType(
+                named("sun.rmi.transport.StreamRemoteCall")
+                    .or(named("sun.rmi.transport.ObjectTable"))
+                    .or(named("sun.rmi.transport.ObjectEndpoint"))));
```
[the remark this answers is not in the document]
Comment (pawelchcki) [reply dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/RmiContextInstrumentation.java]: I think the current approach makes good logical distinctions between components (client, server and propagation). although splitting it into context.client and context.server does make a lot of sense. On it

**Comment by pawelchcki** [comment]
Just noticed I missed few comments - on it.

**Comment by pawelchcki** [comment]
Addressed last remaining comments

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/server/RmiServerContextInstrumentation.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+  @Override
+  public String[] helperClassNames() {
+    return new String[] {
+      "datadog.trace.instrumentation.rmi.context.ContextPayload$InjectAdapter",
+      "datadog.trace.instrumentation.rmi.context.ContextPayload$ExtractAdapter",
+      "datadog.trace.instrumentation.rmi.context.ContextPayload",
+      "datadog.trace.instrumentation.rmi.context.ContextPropagator",
+      packageName + ".ContextDispatcher",
+      packageName + ".ObjectTableAdvice$NoopRemote"
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/server/RmiServerContextInstrumentation.java]: Any time you have to inject an `Advice` class, that's a warning for me...  Can `NoopRemote` go somewhere else?

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ContextPropagator.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+import sun.rmi.transport.TransportConstants;
+
+@Slf4j
+public class ContextPropagator {
+  private static final ObjID ACTIVATOR_ID = new ObjID(ObjID.ACTIVATOR_ID);
+  private static final ObjID DGC_ID = new ObjID(ObjID.DGC_ID);
+  private static final ObjID REGISTRY_ID = new ObjID(ObjID.REGISTRY_ID);
+  public static final ObjID DD_CONTEXT_CALL_ID = new ObjID("Datadog.v1.context_call".hashCode());
+  private static final int CONTEXT_CHECK_CALL_OP_ID = -1;
+  public static final int CONTEXT_PASS_OPERATION_ID = -2;
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ContextPropagator.java]: This class, as well as these constants could use some docs.

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/client/RmiClientAdvice.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+import static datadog.trace.instrumentation.api.AgentTracer.startSpan;
+import static datadog.trace.instrumentation.rmi.client.RmiClientDecorator.DECORATE;
+
+import datadog.trace.api.DDTags;
+import datadog.trace.instrumentation.api.AgentScope;
+import datadog.trace.instrumentation.api.AgentSpan;
+import java.lang.reflect.Method;
+import net.bytebuddy.asm.Advice;
+
+public class RmiClientAdvice {
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/client/RmiClientAdvice.java]: Now that most of the logic is pulled out of the `Advice` classes, they're small enough they can be nested within the instrumentation class.  (This is completely optional and not very consistent as some instrumentation still keeps the advice class separate.)

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/server/ContextDispatcher.java [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+import datadog.trace.instrumentation.api.AgentSpan;
+import datadog.trace.instrumentation.rmi.context.ContextPayload;
+import datadog.trace.instrumentation.rmi.context.ContextPropagator;
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.rmi.Remote;
+import java.rmi.server.RemoteCall;
+import sun.rmi.server.Dispatcher;
+
+public class ContextDispatcher implements Dispatcher {
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/server/ContextDispatcher.java]: Could you add a class comment here?

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ContextPayload.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class ContextPayload implements Serializable {
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ContextPayload.java]: Class comment?

**Comment by pawelchcki** [comment]
Thanks @tylerbenson! Code is ready for next review

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/client/RmiClientAdvice.java (hunk 11 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/client/RmiClientAdvice.java]: You nested some other Advice classes, but not this one.  Was that intentional?

**Comment by pawelchcki** [comment]
Thanks @tylerbenson - added last two fixes. Merging now! :)

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/client/RmiClientDecorator.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+import datadog.trace.agent.decorator.ClientDecorator;
+import datadog.trace.api.DDSpanTypes;
+
+public class RmiClientDecorator extends ClientDecorator {
+  public static final RmiClientDecorator DECORATE = new RmiClientDecorator();
+
+  @Override
+  protected String[] instrumentationNames() {
+    return new String[] {"rmi"};
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/client/RmiClientDecorator.java]: This should include `rmi-client`.

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/server/RmiServerDecorator.java [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+
+import datadog.trace.agent.decorator.ServerDecorator;
+import datadog.trace.api.DDSpanTypes;
+
+public class RmiServerDecorator extends ServerDecorator {
+  public static final RmiServerDecorator DECORATE = new RmiServerDecorator();
+
+  @Override
+  protected String[] instrumentationNames() {
+    return new String[] {"rmi"};
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/server/RmiServerDecorator.java]: This should include `rmi-server`.

File: dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/client/RmiClientDecorator.java [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+  }
+
+  @Override
+  protected String component() {
+    return "rmi-client";
+  }
+
+  @Override
+  protected String service() {
+    return "rmi";
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/client/RmiClientDecorator.java]: Now that I think about it, we probably want this to be null since we expect the remote server to be traced and we don't want the extra hop in the service map.  @pawelchcki Can you submit a separate PR to fix this?

Comment (dougqh) [reply dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/client/RmiClientDecorator.java]: Yes, that probably sense in the usual case.  But isn't that a bit situational?  If they were calling a 3rd party service, then we'd probably prefer it as is.

I think is another example where our notion of peer service is slightly off.

Comment (pawelchcki) [reply dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/client/RmiClientDecorator.java]: If they were calling 3rd party service it would make sense to keep it. I think with RMI though chance of calling 3rd party service is slim :)
